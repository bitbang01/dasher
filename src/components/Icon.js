import React from 'react'
import { Svg } from 'glamorous'
import { oneOfType, number, string } from 'prop-types'

function Icon({ size, ...props }) {
  return (
    <Svg
      xmlns="http://www.w3.org/2000/svg"
      width={size}
      height={size}
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth={2}
      strokeLinecap="round"
      strokeLinejoin="round"
      verticalAlign="middle"
      {...props}
    />
  )
}

Icon.propTypes = {
  size: oneOfType([number, string]),
}

Icon.defaultProps = {
  size: 24,
}

export function IssueIcon(props) {
  return (
    <Icon {...props}>
      <circle cx="12" cy="12" r="10" />
      <line x1="12" y1="8" x2="12" y2="12" />
      <line x1="12" y1="16" x2="12" y2="16" />
    </Icon>
  )
}

export function PullRequestIcon(props) {
  return (
    <Icon {...props}>
      <circle cx="18" cy="18" r="3" />
      <circle cx="6" cy="6" r="3" />
      <path d="M13 6h3a2 2 0 0 1 2 2v7" />
      <line x1="6" y1="9" x2="6" y2="21" />
    </Icon>
  )
}

export function MergeIcon(props) {
  return (
    <Icon {...props}>
      <circle cx="18" cy="18" r="3" />
      <circle cx="6" cy="6" r="3" />
      <path d="M6 21V9a9 9 0 0 0 9 9" />
    </Icon>
  )
}

export function CheckIcon(props) {
  return (
    <Icon {...props}>
      <polyline points="20 6 9 17 4 12" />
    </Icon>
  )
}

export function XIcon(props) {
  return (
    <Icon {...props}>
      <line x1="18" y1="6" x2="6" y2="18" />{' '}
      <line x1="6" y1="6" x2="18" y2="18" />
    </Icon>
  )
}

export function CircleIcon(props) {
  return (
    <Icon {...props}>
      <circle cx="12" cy="12" r="10" />
    </Icon>
  )
}

export function EllipsesIcon(props) {
  return (
    <Icon {...props}>
      <circle cx="12" cy="12" r="1" />
      <circle cx="19" cy="12" r="1" />
      <circle cx="5" cy="12" r="1" />
    </Icon>
  )
}

export default Icon
