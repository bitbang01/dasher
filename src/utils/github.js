import qs from 'qs'

const GITHUB_CLIENT_ID = '799bd70f09884c025750'

export function getGithubAuthUrl({ origin }) {
  const redirectUri = origin + '/callback'

  const queryString = qs.stringify({
    client_id: GITHUB_CLIENT_ID,
    scope: 'repo',
    redirect_uri: redirectUri,
  })

  return `https://github.com/login/oauth/authorize?${queryString}`
}
