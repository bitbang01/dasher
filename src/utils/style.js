import color from 'color-string'
import memoize from 'fast-memoize'

const isNumber = value => typeof value === 'number' && !isNaN(value)

const toUnit = unit => value =>
  isNumber(value) && value !== 0 ? value + unit : value

export const toPx = toUnit('px')

export const toEm = toUnit('em')

export const toMediaQuery = breakpoint =>
  `@media screen and (min-width: ${toEm(breakpoint)})`

export const joinSpacing = (...args) => {
  if (args.length > 4) {
    throw new Error('Too many arguments')
  }

  return args.map(toPx).join(' ')
}

/**
 * Returns the transparent version of the `foreground` color,
 * as if it was overlaid on the `background` color.
 *
 * Inspiration: https://git.io/vpYAv
 *
 * @param {string} foreground
 * @param {string} [background=#fff]
 * @returns {string}
 */
export const toAlpha = memoize(
  (foreground, background = '#fff') => {
    const r = 0
    const g = 1
    const b = 2

    // Convert color string to array (i.e. [r, g, b, a])
    const fgColor = color.get.rgb(foreground)
    const bgColor = color.get.rgb(background)

    // Calculate alpha value
    let alpha = [r, g, b]
      .map(
        channel =>
          (fgColor[channel] - bgColor[channel]) /
          ((0 < fgColor[channel] - bgColor[channel] ? 255 : 0) -
            bgColor[channel]),
      )
      .sort((a, b) => b - a)[0]

    // Keep alpha value between 0 and 1
    alpha = Math.max(Math.min(alpha, 1), 0)

    // Calculate the resulting color
    function processChannel(channel) {
      return 0 === alpha
        ? bgColor[channel]
        : bgColor[channel] + (fgColor[channel] - bgColor[channel]) / alpha
    }

    return color.to.rgb(
      processChannel(r),
      processChannel(g),
      processChannel(b),
      Math.round(alpha * 100) / 100,
    )
  },
  {
    strategy: memoize.strategies.variadic,
  },
)

/**
 * Makes it easier to create glamorous components that
 * accept props to enable/disable styles.
 *
 * Accepts an object where the key is a prop and the value is
 * the styles object that should be applied if that prop is passed.
 * Returns a function which you pass to a glamorousComponentFactory.
 *
 * Source: https://codesandbox.io/s/AGRRMl63 by @kentcdodds
 *
 * @param {Object} styles - An object mapping props to styles
 * @return {Function} - A dynamic styles function
 */
export const propStyles = styles => (props = {}) =>
  Object.keys(props).map(key => {
    if (props[key]) {
      return typeof styles[key] === 'function'
        ? styles[key](props)
        : styles[key]
    }
    return null
  })
