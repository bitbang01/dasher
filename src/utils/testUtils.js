// Named this file `testUtils.js` instead of `test.js` because Jest treats
// `test.js`as a test file
// Source: https://jestjs.io/docs/en/configuration#testmatch-array-string

import { ThemeProvider } from 'emotion-theming'
import React from 'react'
import { render } from 'react-testing-library'
import theme from '../theme'

export const renderWithTheme = children =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
