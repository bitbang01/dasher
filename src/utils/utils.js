export function get(path, object) {
  return path.reduce(
    (acc, property) => (acc && acc[property] ? acc[property] : null),
    object,
  )
}

export function reorder(list, from, to) {
  const listCopy = [...list]
  const [removed] = listCopy.splice(from, 1)
  listCopy.splice(to, 0, removed)
  return listCopy
}
