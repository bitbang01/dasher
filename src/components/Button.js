import { bool, func, oneOf } from 'prop-types'
import styled from 'react-emotion'
import {
  colors,
  focusStyle,
  fontSizes,
  fontWeights,
  lineHeights,
  radii,
  space,
  transition,
} from '../theme'
import { joinSpacing, propStyles, toAlpha } from '../utils/style'
import { cleanElement } from '../utils/utils'

const buttonStyles = {
  primary: {
    color: colors.white,
    backgroundColor: colors.indigo[7],

    ':hover': {
      backgroundColor: colors.indigo[9],
    },

    ':focus': focusStyle,
  },
  secondary: {
    color: toAlpha(colors.gray[9]),
    backgroundColor: 'transparent',

    ':hover': {
      backgroundColor: toAlpha(colors.gray[1]),
    },

    ':focus': focusStyle,
  },
  danger: {
    color: colors.white,
    backgroundColor: colors.red[7],

    ':hover': {
      backgroundColor: colors.red[9],
    },

    ':focus': focusStyle,
  },
  icon: {
    padding: space[1],
    color: colors.gray[6],
    backgroundColor: 'transparent',
    verticalAlign: 'middle',

    ':hover,:focus': {
      backgroundColor: toAlpha(colors.gray[1]),
    },
  },
}

const Button = styled(cleanElement({ type: 'button' }))(
  {
    display: 'inline-block',
    padding: joinSpacing(space[2], space[3]),
    fontFamily: 'inherit',
    fontSize: fontSizes[1],
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.tight,
    textDecoration: 'none',
    border: 0,
    borderRadius: radii[1],
    outline: 0,
    whiteSpace: 'nowrap',
    cursor: 'pointer',
    userSelect: 'none',
    transition: `all ${transition.duration} ${transition.easing}`,
  },
  props => buttonStyles[props.buttonStyle],
  propStyles({
    disabled: {
      opacity: 0.5,
      pointerEvents: 'none',
    },
  }),
)

Button.propTypes = {
  buttonStyle: oneOf(['primary', 'secondary', 'danger', 'icon']),
  disabled: bool,
  onClick: func,
}

Button.defaultProps = {
  buttonStyle: 'primary',
  disabled: false,
  onClick: () => {},
}

export default Button
