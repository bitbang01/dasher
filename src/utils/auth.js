import { gql } from 'apollo-boost'

import { client } from '../components/Apollo'

export const AUTH_TOKEN_KEY = 'auth_token'
export const GITHUB_TOKEN_KEY = 'github_token'

const AUTHENTICATE_MUTATION = gql`
  mutation AuthenticateMutation($githubCode: String!) {
    authenticate(githubCode: $githubCode) {
      token
      githubToken
    }
  }
`

export async function logIn(githubCode) {
  const { data } = await client.mutate({
    mutation: AUTHENTICATE_MUTATION,
    variables: { githubCode },
  })

  setAuthToken(data.authenticate.token)
  setGithubToken(data.authenticate.githubToken)
}

export function logOut() {
  removeAuthToken()
  removeGithubToken()
}

export function isLoggedIn() {
  return Boolean(getAuthToken())
}

export function setAuthToken(token) {
  localStorage.setItem(AUTH_TOKEN_KEY, token)
}

export function getAuthToken() {
  return localStorage.getItem(AUTH_TOKEN_KEY)
}

export function removeAuthToken() {
  localStorage.removeItem(AUTH_TOKEN_KEY)
}

export function setGithubToken(githubToken) {
  localStorage.setItem(GITHUB_TOKEN_KEY, githubToken)
}

export function getGithubToken() {
  return localStorage.getItem(GITHUB_TOKEN_KEY)
}

export function removeGithubToken() {
  localStorage.removeItem(GITHUB_TOKEN_KEY)
}
