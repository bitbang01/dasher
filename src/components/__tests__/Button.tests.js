import React from 'react'
import { fireEvent, renderIntoDocument } from 'react-testing-library'
import { renderWithTheme } from '../../utils/testUtils'
import Button from '../Button'
import { EllipsesIcon } from '../Icon'

it('renders without crashing', () => {
  const { container } = renderWithTheme(<Button>Test Button</Button>)
  expect(container.firstChild).toMatchSnapshot()
})

it('renders when disabled', () => {
  const { container } = renderWithTheme(
    <Button disabled>Disabled Button</Button>,
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders as primary', () => {
  const { container } = renderWithTheme(
    <Button buttonStyle="primary">Primary Button</Button>,
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders as secondary', () => {
  const { container } = renderWithTheme(
    <Button buttonStyle="secondary">Primary Button</Button>,
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders as danger', () => {
  const { container } = renderWithTheme(
    <Button buttonStyle="danger">Primary Button</Button>,
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders as icon', () => {
  const { container } = renderWithTheme(
    <Button buttonStyle="icon">
      <EllipsesIcon />
    </Button>,
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('calls onClick when button is clicked', () => {
  const handleClick = jest.fn()
  const { getByText } = renderIntoDocument(
    <Button onClick={handleClick}>Clickable Button</Button>,
  )
  fireEvent(
    getByText('Clickable Button'),
    new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }),
  )
  expect(handleClick).toHaveBeenCalledTimes(1)
})
