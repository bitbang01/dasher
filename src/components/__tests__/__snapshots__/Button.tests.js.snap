// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as danger 1`] = `
.css-0,
[data-css-0] {
  display: inline-block;
  padding: 8px 12px;
  border: none;
  border-radius: 3px;
  outline: 0;
  font-family: inherit;
  font-size: 14px;
  font-weight: 600;
  line-height: 1.25;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: #ffffff;
  background-color: #f03e3e;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #c92a2a;
}

.css-0:focus,
[data-css-0]:focus {
  box-shadow: 0 0 0 3px rgba(255, 0, 0, 0.34);
}

<button
  class="css-0"
>
  Primary Button
</button>
`;

exports[`renders as primary 1`] = `
.css-0,
[data-css-0] {
  display: inline-block;
  padding: 8px 12px;
  border: none;
  border-radius: 3px;
  outline: 0;
  font-family: inherit;
  font-size: 14px;
  font-weight: 600;
  line-height: 1.25;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: #ffffff;
  background-color: #4263eb;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #364fc7;
}

.css-0:focus,
[data-css-0]:focus {
  box-shadow: 0 0 0 3px rgba(0, 51, 255, 0.43);
}

<button
  class="css-0"
>
  Primary Button
</button>
`;

exports[`renders as secondary 1`] = `
.css-0,
[data-css-0] {
  display: inline-block;
  padding: 8px 12px;
  border: none;
  border-radius: 3px;
  outline: 0;
  font-family: inherit;
  font-size: 14px;
  font-weight: 600;
  line-height: 1.25;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: rgba(0, 5, 9, 0.87);
  background-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: rgba(0, 36, 73, 0.05);
}

.css-0:focus,
[data-css-0]:focus {
  box-shadow: 0 0 0 3px rgba(0, 31, 62, 0.13);
}

<button
  class="css-0"
>
  Primary Button
</button>
`;

exports[`renders at fullWidth 1`] = `
.css-0,
[data-css-0] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  padding: 8px 12px;
  border: none;
  border-radius: 3px;
  outline: 0;
  font-family: inherit;
  font-size: 14px;
  font-weight: 600;
  line-height: 1.25;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: #ffffff;
  background-color: #4263eb;
  width: 100%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #364fc7;
}

.css-0:focus,
[data-css-0]:focus {
  box-shadow: 0 0 0 3px rgba(0, 51, 255, 0.43);
}

<button
  class="css-0"
>
  Full Width Button
</button>
`;

exports[`renders when disabled 1`] = `
.css-0,
[data-css-0] {
  display: inline-block;
  padding: 8px 12px;
  border: none;
  border-radius: 3px;
  outline: 0;
  font-family: inherit;
  font-size: 14px;
  font-weight: 600;
  line-height: 1.25;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: #ffffff;
  background-color: #4263eb;
  opacity: 0.5;
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #364fc7;
}

.css-0:focus,
[data-css-0]:focus {
  box-shadow: 0 0 0 3px rgba(0, 51, 255, 0.43);
}

<button
  disabled=""
  class="css-0"
>
  Disabled Button
</button>
`;

exports[`renders without crashing 1`] = `
.css-0,
[data-css-0] {
  display: inline-block;
  padding: 8px 12px;
  border: none;
  border-radius: 3px;
  outline: 0;
  font-family: inherit;
  font-size: 14px;
  font-weight: 600;
  line-height: 1.25;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  transition: all 150ms ease-in-out;
  color: #ffffff;
  background-color: #4263eb;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: all 150ms ease-in-out;
  -moz-transition: all 150ms ease-in-out;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #364fc7;
}

.css-0:focus,
[data-css-0]:focus {
  box-shadow: 0 0 0 3px rgba(0, 51, 255, 0.43);
}

<button
  class="css-0"
>
  Test Button
</button>
`;
